lib/collection/src/collection/query.rs

```diff
@@ -60,6 +61,83 @@ impl Collection {
         Ok(results.into_iter().next().unwrap())
     }
 
+    /// If the query limit above this value, it will be a subject to undersampling.
+    const SHARD_QUERY_SUBSAMPLING_LIMIT: usize = 128;
+
+    /// Give some more ensurance for undersampling,
+    /// retrieve more points to prevent undersampling errors.
+    /// Errors are still possible, but rare enough to be acceptable compared to
+    /// errors introduced by vector index.
+    const MORE_ENSURANCE_FACTOR: f64 = 1.2;
+
+    /// Creates a copy of requests in case it is possible to apply limit modification
+    /// Returns unchanged requests if limit modification is not applicable.
+    ///
+    /// If there are many independent shards, and we need a very high limit, we can do an optimization.
+    /// Instead of querying all shards with the same limit, we can query each shard with a smaller limit
+    /// and then merge the results. Since shards are independent and data is randomly distributed, we can
+    /// apply probability estimation to make sure we query enough points to get the desired number of results.
+    ///
+    /// Same optimization we already apply on segment level, but here it seems to be even more reliable
+    /// because auto-sharding guarantee random and independent distribution of data.
+    ///
+    /// Unlike segments, however, the cost of re-requesting the data is much higher for shards.
+    /// So we "accept" the risk of not getting enough results.
+    fn modify_shard_query_for_undersampling_limits(
+        batch_request: Arc<Vec<ShardQueryRequest>>,
+        num_shards: usize,
+        is_auto_sharding: bool,
+    ) -> Arc<Vec<ShardQueryRequest>> {
+        if num_shards <= 1 {
+            return batch_request;
+        }
+
+        // Check this parameter inside the function
+        // to ensure it is not omitted in the future.
+        if !is_auto_sharding {
+            return batch_request;
+        }
+
+        let max_limit = batch_request
+            .iter()
+            .map(|req| req.limit + req.offset)
+            .max()
+            .unwrap_or(0);
+
+        if max_limit < Self::SHARD_QUERY_SUBSAMPLING_LIMIT {
+            return batch_request;
+        }
+
+        let mut new_requests = Vec::with_capacity(batch_request.len());
+
+        for request in batch_request.iter() {
+            let mut new_request = request.clone();
+            let request_limit = new_request.limit + new_request.offset;
```

Comment:
Should we not go lower than the offset, to not destabilize results before our offset?

I understand that this wasn't 100% accurate or deterministic before, but I wonder if this might make it worse, amplifying inaccuracies, making the results all over the place.

Reply:
To compensate the accuracy user might increase hnsw_ef, it will have the same effect as higher limits, but won't cause network/IO saturation as bad as limit